lib/tests/test_rewrite.rs

```diff
@@ -66,6 +70,153 @@ fn test_restore_tree() {
+        },
+        hashset! {},
+        RebaseOptions {
+            empty: empty.clone(),
+        },
+    );
+
+    let new_head = match empty {
+        EmptyBehaviour::Keep => {
+            // The commit C isn't empty.
+            let new_commit_c =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_c, &[&commit_bd]);
+            let new_commit_d =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_d, &[&commit_bd]);
+            let new_commit_e =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_e, &[&commit_bd]);
+            let new_commit_f = assert_rebased(
+                rebaser.rebase_next().unwrap(),
+                &commit_f,
+                &[&new_commit_c, &new_commit_d, &new_commit_e],
+            );
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_f])
+        }
+        EmptyBehaviour::AbandonAllEmpty => {
+            // The commit C isn't empty.
+            let new_commit_c =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_c, &[&commit_bd]);
+            // D and E are empty, and F is a clean merge with only one child. Thus, F is
+            // also considered empty.
+            assert_rebase_skipped(rebaser.rebase_next().unwrap(), &commit_d, &commit_bd);
+            assert_rebase_skipped(rebaser.rebase_next().unwrap(), &commit_e, &commit_bd);
+            assert_rebase_skipped(rebaser.rebase_next().unwrap(), &commit_f, &new_commit_c);
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_c])
+        }
+        EmptyBehaviour::AbandonNewlyEmpty => {
+            // The commit C isn't empty.
+            let new_commit_c =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_c, &[&commit_bd]);
+
+            // The changes in D are included in BD, so D is newly empty.
+            assert_rebase_skipped(rebaser.rebase_next().unwrap(), &commit_d, &commit_bd);
+            // E was already empty, so F is a merge commit with C and E as parents.
+            // Although it's empty, we still keep it because we don't want to drop merge
+            // commits.
+            let new_commit_e =
+                assert_rebased(rebaser.rebase_next().unwrap(), &commit_e, &[&commit_bd]);
+            let new_commit_f = assert_rebased(
+                rebaser.rebase_next().unwrap(),
+                &commit_f,
+                &[&new_commit_c, &new_commit_e],
+            );
+            assert_rebased(rebaser.rebase_next().unwrap(), &commit_g, &[&new_commit_f])
+        }
+    };
+
+    assert!(rebaser.rebase_next().unwrap().is_none());
+    assert_eq!(rebaser.rebased().len(), 5);
+
+    assert_eq!(
+        *tx.mut_repo().view().heads(),
+        hashset! {
+            new_head.id().clone(),
+        }
+    );
+}
+
+#[test]
+fn test_rebase_abandon_all_empty() {
+    test_empty_commit_option(EmptyBehaviour::AbandonAllEmpty)
+}
```

Comment:
nit: could be `#[test_case(EmptyBehavior::..)]`?

Reply:
Oh, that looks like exactly what I was looking for, thanks